File: 数据结构和算法/python/mergeSort.py
# 归并排序的实现

def mergeSort(array, p, r):
    if p >= r:
        return
    q = (p + r) // 2
    mergeSort(array, p, q)
    mergeSort(array, q + 1, r)
    merge(array, p, q, r)

    
def merge(array, p, q, r):
    i, j = p, q + 1
    tmp = []
    while i <= q and j <= r:
        if array[i] <= array[j]:
            tmp.append(array[i])
            i += 1
        else:
            tmp.append(array[j])
            j += 1
    
    start, end = i, q
    if j <= r:
        start, end = j, r

    while start <= end:
        tmp.append(array[start])
        start += 1
    
    i = 0
    for i in range(r - p + 1):
        array[p + i] = tmp[i]


if __name__ == "__main__":
    arr = [1, 3, 5, 2, 4, 6, 9, 7, -1]
    mergeSort(arr, 0, len(arr) - 1)
    print (arr)

File: 数据结构和算法/java/MergeSort.java
public class MergeSort{


    public void sort(int[] array,int start ,int end){
        if (start >= end){
            return;
        }
        int middle = (start + end)/2;
        // 拆分
        sort(array, start, middle);
        sort(array, middle + 1, end);
        // 合并
        int[] tmp = new int[end - start + 1];
        int s = start;
        int m = middle + 1;
        int t = 0;
        while(s <= middle && m <= end) {
            if (array[s] >= array[m]){
                tmp[t++] = array[s++];
            } else {
                tmp[t++] = array[m++];
            }
        }
        int x = s;
        int y = middle;
        if (m <= end){
            x = m;
            y = end;
        }
        while (x <= y){
            tmp[t++] = array[x++];
        }
        for (int i = 0; i < tmp.length; i++){
            array[start++] = tmp[i];
        }
    }

    public static void main(String[] args) {
        int[] array = {1, 3, 8, 6, 4, 2, -1, 8};
        MergeSort mergesort = new MergeSort();
        mergesort.sort(array, 0, array.length - 1);
        for (int num : array){
            System.out.println(num);
        }
    }

}